# Практическая работа №2: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 1304 Заика Тимофей. Вариант №35.

## Цель работы

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа, привить навык использования систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения

### Понятие аддитивной цепочки

Аддитивной цепочкой натурального числа $n$ называется последовательность натуральных чисел<br> $$ 1=a_0,a_1,a_2,\dots,a_m=n,\quad   a_i=a_j+a_k, \quad \forall k \leq j < i$$ Число $r$ называется длиной цепочки.<br>Длина кратчайшей цепочки для заданного $n$ обозначается $l(n) = r$<br>
Цепочка называется звёздной, если каждый каждый её элемент получается следующим образом:<br>
$$a_i=a_{i-1}+a_k\quad \forall k<i$$
Справедливы некоторые оценки для $l(n)$:<br>
$$l(n) \leq \lambda(n) +\nu(n) - 1$$
$$l(n) \geq \lceil \log_{2}(n) \rceil$$

### Алгоритм Брауэра
Для $n \in \mathbb{N}$ при заданном $k \in \mathbb{N}$ можно построить цепочку Брауэра с помощью рекуррентной формулы:<br>
$$ B_{k}(n) =
  \begin{cases}
    1,2,3,\dots,2^{k} - 1       & \quad n < 2^{k}\\
    B_{k}(q),2q,4q,8q,\dots,2^{k}q,n  & \quad n \ge 2^{k}, q = \lfloor\frac{n}{2^{k}}\rfloor \\
  \end{cases}
$$
Данная цепочка будет иметь длину:
$$ l_{B}(n) = j(k + 1) + 2^{k} - 2, $$ при условии, что $jk \le \lambda(n) \lt (j + 1)k$

Длина будет минимизирована  для больших $n$, если положить, что $k = \lambda\lambda(n) - 2\lambda\lambda\lambda(n)$

Сам алгоритм:

1. Задается некий фиксируемый $k$ для рассматривемого $n$. Выполняется вычисление "вспомогательных чисел":
$$d = 2^{k}, q_1 = \left[\frac{n}{d}\right], \quad r_1 = n \, \mathrm{mod} \, d \Rightarrow n = q_1 d + r_1 \quad(0 \le r_1 \lt d)$$
$$q_2 = \left[\frac{q_1}{d}\right], \quad r_2 = q_1 \, \mathrm{mod} \, d \Rightarrow q_1 = q_2 d + r_2 \quad(0 \le r_2 \lt d)$$

2. Данная процедура продолжается до тех пор, пока не появится такое $q_s \lt d$. Следовательно $q_{s-1} = q_s d + r_s$

3. Таким образом, n имеет вид:
$$n = 2^{k}q_1 + r_1 = 2^{k}\left(2^{k}q_2 + r_2\right) + r_1 = \dots = \\ 2^{k}\left(2^{k} \left(\dots\left(2^{k}q_s + r_s\right)\dots\right) + r_2\right) + r_1 $$

### Алгоритм дробления векторов
Вектором индексов называется последовательность $r_1,r_2,\dots,r_{m-1}$, где $$r_i=\{z:1\leq z\leq i\}, \quad a_i=a_{i-1}+a_{r_{i-1}},\quad2\leq i\leq m-1.$$
Зададим два вектора индексов $r=\{r_i\}_{i=1}^{m-1}$ и $\widetilde r=\{\widetilde r_i\}_{i=1}^{m-1}$. Тогда $r \succ \widetilde r$, если $r_1=\widetilde r_1,r_2=\widetilde r_2,\dots,r_{m-1}=\widetilde r_{m-1}$, а $r_m>\widetilde r_m$.<br>

Пусть задано $n \in \mathbb{N}$. Необходимо найти минимальную звездную цепочку, такую , что $a_m = n$.

Рассмотрим вектор индексов вида $$\left\{r_1,r_2,r_3,\dots,r_q\right\} \cup \left\{\rho_{q + 1},\rho_{q + 2},\rho_{q + 3},\dots,\rho_{m}\right\} $$ 
Назовем левую часть фиксированной, а правую - меняющейся.

$$a_{min} = a_{q+1} + m - q, \quad\text{при} \left\{r_1,r_2,r_3,\dots,r_q\right\} \cup \left\{1,1,1,\dots,1\right\} \\ 
a_{max} = a_{q+1} \cdot 2^{m - q}, \quad\text{при} \left\{r_1,r_2,r_3,\dots,r_q\right\} \cup \left\{q + 1,q + 2,q + 3,\dots,m\right\}$$

Сам алгоритм:
1. Запускаем внешний цикл по длинам цепочек: $\underline{l}(n) \le m \le \overline{l}(n)$.<br>
    Выбираем число $q \in \mathbb{N}$. Пусть $q = \frac{m}{2}$
2. Запускаем внутренний цекл перебора всех $\left\{r_1,r_2,r_3,\dots,r_q\right\}$ ($q!$ шагов). <br>
    На каждом шаге при фиксированной части вычисляем $a_{min}$ и $a_{max}$<br>
    1) Если $a_m = n$ - задача решена<br>
    2) Если $n \notin \left[a_{min},a_{max}\right]$, то переходим к следующему набору $\left\{r_1,r_2,r_3,\dots,r_q\right\}$<br>
    3) Если $n \in \left[a_{min},a_{max}\right]$, то организуем внутренний цикл перебора меняющейся части $\left\{\rho_{q + 1},\rho_{q + 2},\rho_{q + 3},\dots,\rho_{m}\right\}$. Таких наборов $\frac{m!}{q!}$ штук:<br>
    3.1) Если обнаруживается $a_m = n$ - задача решена.<br>
    3.2) Если в цикле таких векторов не оказалось, то переходим к следующей (по введенной упорядоченности) фиксированной части $\left\{r_1,r_2,r_3,\dots,r_q\right\}$
3. Если все наборы фиксированной длины исчерпаны, то увеличиваем их длину во внешнем цикле.

### Гипотеза Шольца-Брауэра
Гипотеза заключается в следующем: 
$$l(2^{n} - 1) \le l(n) + n - 1$$
1. Гипотеза доказана для звездных цепочек: $l^{*}(2^{n} - 1) \le l^{*}(n) + n - 1$
2. Гипотеза справедлива для всех $n \lt 5784689$
3. Равенство выполняется для всех $1 \le n \le 64$

## Постановка задачи

Реализовать точные и приближенные алгоритмы нахождения минимальных аддитвных цепочек с использованием системы компьютерной математики <i>SageMath</i>, провести анализ алгоритмов. Полученные результаты содержательно проинтерпертировать.

## Выполнение работы

### 1. Алгоритм Брауэра

Создадим функцию brauer_algorythm.<br>
На вход принимает число $n$, для которого необходимо найти приближенную аддитивную цепочку, и параметр $k$.<br>
Возвращает кортеж, первый элементо которого массив вспомогательных чисел, а второй - массив с элементами аддитивной цепочки.<br>

In [42]:
def brauer_algorythm(n, k):
    d = pow(2, k)
    r_array = list()
    q_array = list()
    q_current = n
    
    while q_current >= d:
        q_tmp = q_current // d
        q_array.append(q_tmp)
        r_tmp = q_current % d
        r_array.append(r_tmp)
        q_current = q_tmp
        
    additive_chain = list()
    helping_numbers = list()
    
    for i in range(1, d):
        helping_numbers.append(i)
    
    part_of_chain = q_array[len(q_array) - 1]
    
    for i in range(len(q_array) - 1, -1, -1):
        for j in range(k):
            part_of_chain *= 2
            additive_chain.append(part_of_chain)
        part_of_chain += r_array[i]
        if r_array[i] != 0:
            additive_chain.append(part_of_chain)
            
    while additive_chain[0] < d:
        additive_chain.pop(0)
            
    return (helping_numbers, additive_chain)

Создадим функции для печати результата работы алгоритма, а также для его сравнения с минимальными аддитивными цепочками в виде таблицы.

In [43]:
def print_result_of_brauer(result):
    helping_numbers, additive_chain = result[0], result[1]
    print("Вспомогательные числа: ", helping_numbers)
    print("Аддитивная цепочка: ", additive_chain)
    print("Длина аддитивной цепочки: ", len(additive_chain), "\n")
    
def print_compare_brauer_vs_min_chains_in_table(array_of_n, min_chains):
    print("| n | k | Длина цепочки Брауэра | Аддитивная цепочка Брауэера | Длина минимальной цепочки | Минимальная цепочка |")
    print('|---|---|-----------------------|-----------------------------|---------------------------|---------------------|')
    for n in array_of_n:
        for k in range(2, 5):
            result = brauer_algorythm(n, k)
            additive_chain = result[0]
            additive_chain += result[1]
            print('|', n, '|', k,' |', len(additive_chain), '|', additive_chain, '|', len(min_chains[n]), '|', min_chains[n], '|')

Зададим произвольные $n$ и минимальные аддитивные цепочки для них.

In [44]:
array_of_n = [232,511,723,1024,1345]
min_chains = {
    232: [1,2,3,6,7,14,28,29,58,116,232],
    511: [1,2,3,6,12,15,30,60,120,240,255,510,511],
    723: [1,2,3,5,10,20,40,45,90,180,360,720,723],
    1024: [1,2,4,8,16,32,64,128,256,512,1024],
    1345: [1,2,4,5,10,20,21,42,84,168,336,672,1344,1345]
}

Протестируем алгоритм на различных $n$ и $k$:

In [45]:
for n in array_of_n:
    for k in range(2, 5):
        print(f"n = {n}, k = {k}")
        print_result_of_brauer(brauer_algorythm(n, k))

n = 232, k = 2
Вспомогательные числа:  [1, 2, 3]
Аддитивная цепочка:  [6, 12, 14, 28, 56, 58, 116, 232]
Длина аддитивной цепочки:  8 

n = 232, k = 3
Вспомогательные числа:  [1, 2, 3, 4, 5, 6, 7]
Аддитивная цепочка:  [12, 24, 29, 58, 116, 232]
Длина аддитивной цепочки:  6 

n = 232, k = 4
Вспомогательные числа:  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
Аддитивная цепочка:  [28, 56, 112, 224, 232]
Длина аддитивной цепочки:  5 

n = 511, k = 2
Вспомогательные числа:  [1, 2, 3]
Аддитивная цепочка:  [4, 7, 14, 28, 31, 62, 124, 127, 254, 508, 511]
Длина аддитивной цепочки:  11 

n = 511, k = 3
Вспомогательные числа:  [1, 2, 3, 4, 5, 6, 7]
Аддитивная цепочка:  [14, 28, 56, 63, 126, 252, 504, 511]
Длина аддитивной цепочки:  8 

n = 511, k = 4
Вспомогательные числа:  [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
Аддитивная цепочка:  [16, 31, 62, 124, 248, 496, 511]
Длина аддитивной цепочки:  7 

n = 723, k = 2
Вспомогательные числа:  [1, 2, 3]
Аддитивная цепочка:  [4, 8, 11, 2

Приведем таблицу для разлиных $n$ и $k$ и сравним результат работы алгоритма с минимальными аддитивными цепочками для заданных $n$:

| n | k | Длина цепочки Брауэра | Аддитивная цепочка Брауэера | Длина минимальной цепочки | Минимальная цепочка |
|---|---|-----------------------|-----------------------------|---------------------------|---------------------|
| 232 | 2  | 11 | [1, 2, 3, 6, 12, 14, 28, 56, 58, 116, 232] | 11 | [1, 2, 3, 6, 7, 14, 28, 29, 58, 116, 232] |
| 232 | 3  | 13 | [1, 2, 3, 4, 5, 6, 7, 12, 24, 29, 58, 116, 232] | 11 | [1, 2, 3, 6, 7, 14, 28, 29, 58, 116, 232] |
| 232 | 4  | 20 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 28, 56, 112, 224, 232] | 11 | [1, 2, 3, 6, 7, 14, 28, 29, 58, 116, 232] |
| 511 | 2  | 14 | [1, 2, 3, 4, 7, 14, 28, 31, 62, 124, 127, 254, 508, 511] | 13 | [1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 511] |
| 511 | 3  | 15 | [1, 2, 3, 4, 5, 6, 7, 14, 28, 56, 63, 126, 252, 504, 511] | 13 | [1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 511] |
| 511 | 4  | 22 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 31, 62, 124, 248, 496, 511] | 13 | [1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 511] |
| 723 | 2  | 14 | [1, 2, 3, 4, 8, 11, 22, 44, 45, 90, 180, 360, 720, 723] | 13 | [1, 2, 3, 5, 10, 20, 40, 45, 90, 180, 360, 720, 723] |
| 723 | 3  | 17 | [1, 2, 3, 4, 5, 6, 7, 8, 11, 22, 44, 88, 90, 180, 360, 720, 723] | 13 | [1, 2, 3, 5, 10, 20, 40, 45, 90, 180, 360, 720, 723] |
| 723 | 4  | 23 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 32, 45, 90, 180, 360, 720, 723] | 13 | [1, 2, 3, 5, 10, 20, 40, 45, 90, 180, 360, 720, 723] |
| 1024 | 2  | 12 | [1, 2, 3, 4, 8, 16, 32, 64, 128, 256, 512, 1024] | 11 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] |
| 1024 | 3  | 15 | [1, 2, 3, 4, 5, 6, 7, 8, 16, 32, 64, 128, 256, 512, 1024] | 11 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] |
| 1024 | 4  | 22 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 32, 64, 128, 256, 512, 1024] | 11 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] |
| 1345 | 2  | 15 | [1, 2, 3, 4, 5, 10, 20, 21, 42, 84, 168, 336, 672, 1344, 1345] | 14 | [1, 2, 4, 5, 10, 20, 21, 42, 84, 168, 336, 672, 1344, 1345] |
| 1345 | 3  | 17 | [1, 2, 3, 4, 5, 6, 7, 8, 16, 21, 42, 84, 168, 336, 672, 1344, 1345] | 14 | [1, 2, 4, 5, 10, 20, 21, 42, 84, 168, 336, 672, 1344, 1345] |
| 1345 | 4  | 24 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 20, 40, 80, 84, 168, 336, 672, 1344, 1345] | 14 | [1, 2, 4, 5, 10, 20, 21, 42, 84, 168, 336, 672, 1344, 1345] |

### Вывод

Длины аддитивных цепочек, полученные с помощью алгоритма Брауэра, могут как совпадать, так и быть больше чем длины минимальных аддитивных цепочек. Это зависит как от самого числа, так и от подобранного параметра $k$. Алгоритм выдает приближенный ответ за малое время работы.

### 2. Алгоритм дробления вектора индексов

Для оценки длины аддитивных цепочек будем использовать следующие неравенства:
$$l(n) \le \lambda(n) + \upsilon(n) - 1\\
l(n) \ge \lambda(n) + \log_{2}(\upsilon(n))$$
Разработаем следующие функции для данных оценок.

In [62]:
def min_len(n):
    return int(log(n,2)) + int(log(list(bin(n)).count('1'),2))

def max_len(n):
    return int(log(n,2)) + list(bin(n)).count('1') - 1

Далее реализуем алгоритм дробления векторов. Для этого напишем вспомогательные функции <i>get_chain_from_index_vector</i>, которая возвращает аддитивную цепочку по вектору индексов, и <i>go_next_part</i>, которая возвращает следующий относительно данного вектор индексов. Сам алгоритм реализован в функции <i>subdivision_of_the_index_vector</i>.

In [63]:
import time


def get_chain_from_index_vector(index_vector):
    chain = [1]
    
    for idx in index_vector:
        chain.append(chain[-1] + chain[idx - 1])
        
    return chain

def go_next_part(cur_part, q = 0):
    if cur_part == [1 for _ in range(len(cur_part))]:
        return []
    
    for i in range(len(cur_part) - 1, -1, -1):
        if cur_part[i] > 1:
            cur_part[i] -= 1
            break
        elif cur_part[i] == 1:
            cur_part[i] = i + q + 1
    
    return cur_part

def subdivision_of_the_index_vector(n):
    result = []
    lower_estimate = min_len(n)
    high_estimate = max_len(n)
    
    if n < 1:
        return []
    if n == 1:
        return [1]
    
    for m in range(lower_estimate, high_estimate + 1):
        q = m // 2 + 1
        index_vector = [i for i in range(1, m + 1)]
        changeable_indexes = index_vector[:q]
        immutable_indexes = index_vector[q:]
        while (len(changeable_indexes)):
            current_chain = get_chain_from_index_vector(changeable_indexes + immutable_indexes)
            a_min = current_chain[q] + m - q
            a_max = current_chain[q] * 2 ** (m - q)
            if current_chain[-1] == n:
                return current_chain
            elif n < a_min or n > a_max:
                changeable_indexes = go_next_part(changeable_indexes)
            else:
                while (len(immutable_indexes)):
                    current_chain = get_chain_from_index_vector(changeable_indexes + immutable_indexes)
                    if n == current_chain[-1]:
                        return current_chain
                    immutable_indexes = go_next_part(immutable_indexes, q)
                changeable_indexes = go_next_part(changeable_indexes)
                immutable_indexes = index_vector[q:]

Представим работу алгоритма на пяти $n>1000$, а именно при $n = 1003, 1005, 1010, 1011, 1026$.

In [ ]:
def test_subdivision_of_the_index_vector_algorythm():
    n_array = [1003, 1005, 1010, 1011, 1026]
    
    for n in n_array:
        start = time.time()
        chain = subdivision_of_the_index_vector(n)
        end = time.time() - start
        print('|', n, '|', len(chain), '|', chain, '|', round(end, 3) , ' секунд |')

test_subdivision_of_the_index_vector_algorythm()

Результат оформим в виде таблицы.

| n    | Длина цепочки | Цепочка                                                         | Время работы |
|------|---------------|-----------------------------------------------------------------|--------------|
| 1003 | 14 | [1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 201, 401, 802, 1003] | 483.255  секунд |
| 1005 | 14 | [1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 201, 402, 804, 1005] | 486.751  секунд |
| 1010 | 14 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 320, 336, 672, 1008, 1010] | 489.939  секунд |
| 1011 | 14 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 320, 336, 337, 674, 1011] | 478.576  секунд |
| 1026 | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1026] | 0.001  секунд |

### Вывод

Алгоритм позволяет получить минимальную звездную цепочку за счет полного перебора, в связи с чем имеет долгое время работы. Как видно из таблицы, по сравнению с алгоритмом Брауэра, алгоритм дробления вектора индексов является более ресурсозатратным, но в свою очередь гарантированно строит кратчайшую звездную цепочку. 

### 3. Гипотеза Шольца-Брауэра

Проверим гипотезу Шольца-Брауэра для $1 \le n \le 12$, которая гласит, что для этих чисел должно выполняться равенство: 
$$l(2^{n} - 1) = l(n) + n - 1$$

Исходя из изложенных выше теоритеческих положений, создадим функцию <i>subdivision_of_the_index_vector_for_hypothesis</i>, которая будет перебирать только длины, равные $l(n) + n - 1$.

In [66]:
def subdivision_of_the_index_vector_for_hypothesis(n, m):
    result = []
    
    if n < 1:
        return []
    if n == 1:
        return [1]
    
    m = m - 1
    q = m // 2 + 1
    index_vector = [i for i in range(1, m + 1)]
    changeable_indexes = index_vector[:q]
    immutable_indexes = index_vector[q:]
    while (len(changeable_indexes)):
        current_chain = get_chain_from_index_vector(changeable_indexes + immutable_indexes)
        a_min = current_chain[q] + m - q
        a_max = current_chain[q] * 2 ** (m - q)
        if current_chain[-1] == n:
            return current_chain
        elif n < a_min or n > a_max:
            changeable_indexes = go_next_part(changeable_indexes)
        else:
            while (len(immutable_indexes)):
                current_chain = get_chain_from_index_vector(changeable_indexes + immutable_indexes)
                if n == current_chain[-1]:
                    return current_chain
                immutable_indexes = go_next_part(immutable_indexes, q)
            changeable_indexes = go_next_part(changeable_indexes)
            immutable_indexes = index_vector[q:]

Создадим функцию, проверяющую гипотезу для $1 \le n \le 12$.

In [ ]:
def sholz_brauer(left, right):
    for i in range(left, right + 1):
        right_part_of_equal = len(subdivision_of_the_index_vector(i)) + i - 1
        left_part_of_equal = len(subdivision_of_the_index_vector_for_hypothesis(2^i - 1, right_part_of_equal))
        
        if left_part_of_equal > right_part_of_equal:
            print('|', i, '|', left_part_of_equal, '|', right_part_of_equal, '|', "Неверно", '|')
            break
        print('|', i, '|', left_part_of_equal, '|', right_part_of_equal, '|', "Верно", '|')
        
sholz_brauer(1, 12)

Результат представим в виде таблицы.

|  n | $$l(2^{n}-1)$$ | $$l(n) + n - 1$$ |    Результат   |
|----|--------------|----------------|----------------|
| 1 | 1 | 1 | Верно |
| 2 | 3 | 3 | Верно |
| 3 | 5 | 5 | Верно |
| 4 | 6 | 6 | Верно |
| 5 | 8 | 8 | Верно |
| 6 | 9 | 9 | Верно |
| 7 | 11 | 11 | Верно |
| 8 | 11 | 11 | Верно |
| 9 | 13 | 13 | Верно |
| 10 | 14 | 14 | Верно |
| 11 | 16 | 16 | Верно |
| 12 | 16 | 16 | Верно |

### Вывод

Таким образом, проверена корректность гипотезы Шольца-Брауэра для $1 \le n \le 12$. Для ускорения проверки модифицирован алгоритм дробления вектора индексов на основании следствия из теоретических положений.

## Выводы



Сформировано представление об аддитивных цепочках, вырабатано умение составлять и применять алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа. Привит навык использования системы компьютерной математики <i>SageMath</i> для реализации алгоритмов. Реализованы и сравнены алгоритмы Брауэра и дробления вектора индексов. Проверена гипотеза Шольца-Брауэра для $1 \le n \le 12$. Установлено, что не существует эффективного алгоритма для построения минимальной аддитивной цепочки, который бы достигал точного результата за малое количество времени.